### ISSUE TITLE
Discuss: Should variables be detected and highlighted?

### ISSUE THREAD

**Issue by johnstray** [highlightjs/highlight.js#1378]
Taking the following code as an example, the php variable is not detected as a variable when it resides as a function argument or inside any bracketed syntax:

```php
$var_is_detected = ''; # < This is detected as a variable

function my_function ( $var_not_detected, $another_var_not_detected ) { 
    # These arguments are NOT detected as variables...
 }

# The value of 'id' is not detected as a variable...
$working_variable = array ( 'id' => $var_not_detected );
```

Is Detected = variable is highlighted
Not Detected = variable NOT highlighted

**Comment by isagalaev** [comment]
Currently we don't highlight variables in PHP (if they are highlighted in your example it's likely caused by a wrongly detected different language). This was done for consistency with many other languages where variables aren't highlighted either (Java, C, Ruby, Python, JavaScript, etc.)

Given the amount of controversy it is causing we may revisit this decision though… It's very subjective :-)

**Comment by yyyc514** [comment]
@marcoscaceres @egor-rogov 

I sounds like before we were philosophically against this.  Is that still true?  If so I will close this as "wontfix" but if not I'll mark it as a feature request and help needed and we'd be open to PRs.

I'm not sure I completely buy the consistency argument as if I spent an afternoon I think I can find already find a lot of small inconsistencies already between languages and what they support.  It seems if our long term vision is truly "let the best people maintain the grammars" (which has a lot of benefits) then really this would become more of a community by community decision.  

eg, Maybe PHP people like their variables highlighted, maybe XYZ people do not.

**Comment by Rarst** [comment]
See #1142 for some context/discussion at the time amount of highlight dropped for PHP.

In my personal opinion the amount of highlight for PHP fell to... pointless at that point.

**Comment by yyyc514** [comment]
Quoting the original maintainer:

> First, we just want to be consistent across languages. Most highlighters don't highlight local variables in other languages simply because they're not syntactically distinguishable. In PHP (and, say, Perl) they're easy to highlight because of the $ sign. But "because we can" is not reason enough.

> Second, one of our design guidelines is to keep the bulk of code plain and only highlight things that are semantically significant, and highlighting them will add to code readability (all of this highly subjective and fuzzy, of course). Thus, we don't highlight many things that others do: UPPERCASE_CONSTANTS, CamelCaseNames, .method_calls().

But since then time has moved on, so perhaps time to revisit these things...

**Comment by egor-rogov** [comment]
I don't think we have some strict policy nowadays. This is really subjective matter.
At the same time I don't think we should change existing behaviour.

I can see two possible solutions.
The first one is to clone php into separate repo and then adjust it to the taste of its maintainer. I'm not sure it's a very good idea, but probably we'll find it possible?

Second is to come up with some general approach to pass parameters into languages. This would be nice in cases like this.

**Comment by yyyc514** [comment]
> At the same time I don't think we should change existing behaviour.

If it's pretty trivial... why not? Is consistency still the thought?

> The first one is to clone php into separate repo and then adjust it to the taste of its maintainer.

I thought long-term that was the desired goal to reduce stress on the maintainers of the core parser project? 

Consistency is a nice goal, but also "nicest styling possible per language" is another nice goal.  And who would best define that but users/readers of that language itself? :-)

> Second is to come up with some general approach to pass parameters into languages. This would be nice in cases like this.

You mean allow people to extend the grammar with tiny grammar tweaks or alternations?  Do you see a lot of people actually do that?  I think the external PHP repository that does what the PHP people want just "drop it in" seems to be the more likely outcome.

A little bit of this does get at "who are the users of highlight.js"... a general purpose website who wants "consistent highlighting of all code" might be a very different audience than someone who wants the "best highlighting per language".  Those are very different goals,  hard to say which is better or worse.

IF the goal is user-maintained grammars long-term though I think consistency is going to be impossible to maintain (just too many different languages, too many different syntaxes, etc) so "the very best for a particular language" would seem to be the goal that would win, by default.

**Comment by yyyc514** [comment]
Removing `PHP` from the topic here and making this the general purpose topic for discussion so I can file other issue as dups and point them here.

**Comment by yyyc514** [comment]
In another thread I commented on the same type of request (but regarding highlighting the names of calling functions):

> To push this forward for Javascript (or any other language at the moment) the best way would be to find a few people interested in maintaining the Javascript grammar and then fork that into it's own repo and make those improvements there.

> Then we could discuss the value in that new repo becoming the defacto "JS" that's bundled by default with Highlight.js. I think that's how these types of discussions are going to move forward in a meaningful way - by people to show they are willing to put in the effort to maintain these things. That's currently the problem in core. Too many grammars, too little time and maintainers. So you can imagine the hesitancy to add even more features that will break, require maintenance, etc...

> At the very least people would have a choice and could use the built-in one or decide to use the enhanced version and all it offers.

I think that same advice applies here regarding PHP.

**Comment by egor-rogov** [comment]
> > At the same time I don't think we should change existing behaviour.
> 
> If it's pretty trivial... why not? Is consistency still the thought?

Why yes?
It's not about consistency. You're right, we can hardly maintain it across repos and across hundreds of languages.
But all we have now is one opinion that variables should be highlighted. I don't think it's a valid reason to change things (at least, not in the core).

As long as it is just two options (highlight variables or not), it looks simpler to have something like a parameter inside a single grammar. And if we speak of lots of deviating improvements, then yes, another repo with another grammar is more appropriate.

Anyway, I have no objections if someone is willing to maintain an alternative repo for PHP or other language. The only question is what to do with aliases?

**Comment by yyyc514** [comment]
> The only question is what to do with aliases?

Not sure I follow?  I think at some point we'd "package" or "link to" the "official" community grammar... OR we'd let people choose when building the package... so someone would decide if "php" meant the one still in core or the 'community' one...  And all aliases would then refer to whatever was the version plugged in.

Or am I missing something?

**Comment by yyyc514** [comment]
> But all we have now is one opinion that variables should be highlighted.

Evidently at the time we had LOTS of opinions from some vocal PHP people who hated the "improvement" to not highlight variables... so you could imagine in community hands that feature would come back pretty fast. ;-)

**Comment by yyyc514** [comment]
>it looks simpler to have something like a parameter inside a single grammar.

Or just use the CSS rules?

**Comment by yyyc514** [comment]
It sounds like we've sort of resolve this I think:

- There is _some_ value in core itself staying consistent (while there still is a core for grammars).
- IE, if we were to add variables to one we really should add to all (and that's not feasible, time wise)
- There is also some value in legacy also (ie, let's not change it without a good reason).
- So therefore: will not accept PRs against existing languages that add variable highlighting.
- But if community maintainers decide to diverge from this with their own grammars, more power to them - so long as they are the ones maintaining it?

Is that a fair summary?  And would this same summary also apply to similar feature requests:

- Desire to highlight every little punctuation mark, etc

---

I think the issue of making parts of the grammar process configurable would be a larger discussion and could perhaps be moved into a new fresh thread to discuss ideas for how we might go about even if we wanted to?

**Comment by egor-rogov** [comment]
> > The only question is what to do with aliases?
> 
> Not sure I follow? I think at some point we'd "package" or "link to" the "official" community grammar... OR we'd let people choose when building the package... so someone would decide if "php" meant the one still in core or the 'community' one... And all aliases would then refer to whatever was the version plugged in

What if you load two versions of the same language? Say, in-core one from a bungle, and then an alternate.
Probably it's not a problem at all, and definitely we can do something about it, so nevermind.

**Comment by yyyc514** [comment]
Oh I'd say you'd let the latest one override the previous one... that way someone could start with a "default" build if they wanted and then overlay languages on top of it, even if they happened to be built in the existing bundle.

**Comment by egor-rogov** [comment]
> So therefore: will not accept PRs against existing languages that add variable highlighting.

Unless there are lots of people telling us we do it wrong, I guess.

>  But if community maintainers decide to diverge from this with their own grammars, more power to them - so long as they are the ones maintaining it?

Yep.

@marcoscaceres do you agree?

**Comment by egor-rogov** [comment]
> I think the issue of making parts of the grammar process configurable would be a larger discussion and could perhaps be moved into a new fresh thread to discuss ideas for how we might go about even if we wanted to?

I'm not sure we want, so let start such thread when we have reasons for that.

**Comment by yyyc514** [comment]
> I'm not sure we want, so let start such thread when we have reasons for that.

Ok, so no thread yet. :-)

**Comment by yyyc514** [comment]
@egor-rogov Ok I was kind of 50/50 on this before but after reading thru MANY of our grammars we allow variables to be highlighted in a *great many of them*.  Just a small list:

- ISBL
- Powershell
- Awk
- Julia
- Kotlin
- Apache
- Bash
- Elixir
- Ini
- Makefile
- Nginx
- Puppet
- RouterOS
- SCSS

**And that's not even a COMPLETE list.**

That being the case, if someone made a PR adding this support to PHP I can't think of a good reason we should reject it.

**Comment by ChristianIvicevic** [comment]
Blatantly going to copy/paste my comment from the other ticket after @yyyc514 pointed out that it would have been better to post here in the first place.

TL;DR: I am writing a lot of Typescript lately and coming from Markdown to Asciidoctor (which uses highlight.js) I am really missing the contextual parsing and highlighting of variable names, function invocations and possibly types (at some point).

---

For training material and a lot of code that I am writing, I am in the process of migrating over from Markdown to AsciiDoc. Missing some of the highlighting I had previously is really noticeable as you can see in this example:

<img width="538" alt="Screenshot 2019-10-12 at 19 28 42" src="https://user-images.githubusercontent.com/45569997/66705407-86588880-ed26-11e9-879a-eaf713756466.png">

<img width="567" alt="Screenshot 2019-10-12 at 19 29 06" src="https://user-images.githubusercontent.com/45569997/66705409-95d7d180-ed26-11e9-9f3f-1cf41efd824b.png">

The same code snippet is more difficult to read (which is highly subjective of course) and I personally appreciate the distinction of classes/types from object members etc. which is helpful for beginners to distinguish various code elements.

**Comment by yyyc514** [comment]
@ChristianIvicevic In that example how would we know Client and Project are classes?

**Comment by ChristianIvicevic** [comment]
I do believe parsing 3rd party classes is one of the most difficult tasks when working with Typescript source files. This is due to their nature in possibly being intertwined with various syntactical elements and decorations.

Really naive example: assuming everything after the colon is a type you might run into a literal union `doSomething(arg: "div" | "span")` and all of a sudden you run into a scenario wherein you expected a type but got a literal value.

Another example is with regards to the generics. As you can see, `T` and `K` are contextually picked up being types as well even though they are referred to on multiple occasions in different contexts.

My suggestion is to analyze first whether some more naive features can be implemented to improve highlighting and ignoring more complex cases in the beginning:

* Anything of the form `foo(...)` or `bar.foo(...)` that seems to be a function invocation can be highlighted (in my screenshot the `map` method)
* Parameters inside a function declaration, i.e. in the scope of Typescript anything of the form `name: Type` inside parens such as `entity`, `obj` or `propertyNames` without highlighting their next occurences again
*

**Comment by yyyc514** [comment]
> Before going more into detail, I'd like to ask whether it'd be useful to separate out this very specific discussion on the topic of Typescript into its own issue since the original issue was created based on PHP.

Sure, thanks for asking that. :-)  Can you pull this last bit over to a new thread then?

**Comment by yyyc514** [comment]
Closing this as I think the discussion seems to be over (for now).  My personal conclusion is posted above.  We can revisit this (if necessary) whenever this matter actually comes up again in a PR.
